### PR TITLE
InvalidLinkBear: Improve request handling

### PR DESCRIPTION
**Pull request by SanketDG**
For bad connections, this commit will ping google.com everytime
it fails to GET a request. If this fails, then that means the
system has an unstable internet connection.

Fixes https://github.com/coala-analyzer/coala-bears/issues/184

### CODE REVIEW

File: bears/general/InvalidLinkBear.py [hunk 1]
```diff
@@ -33,14 +44,15 @@ def find_links_in_file(file):
...
+                code = self.get_status_code_checked(link)
                 yield line_number + 1, link, code
 
     def run(self, filename, file):
         '''
         Yields results for all invalid links in a file.
         '''
-        for line_number, link, code in InvalidLinkBear.find_links_in_file(file):
+        for line_number, link, code in (InvalidLinkBear.find_links_in_file(
+                                        self, file)):
```
Comment (abhsag24) [inline bears/general/InvalidLinkBear.py]: i guess using `self.find_links_in_file(file)` would save you a line since its nor longer a staticmethod

Comment (SanketDG) [inline bears/general/InvalidLinkBear.py]: ^yeah, I wanted to do it in another commit.

Comment (abhsag24) [inline bears/general/InvalidLinkBear.py]: like a newcomer issue?

Comment (SanketDG) [inline bears/general/InvalidLinkBear.py]: No in this PR, in a separate commit, since it is not associated with the current commit anyway.

Comment (abhsag24) [inline bears/general/InvalidLinkBear.py]: so you had to change the implementation of `find_links_in_file` right? that does correspond to how it is called from within other functions, so i don't see how changing `InvalidLinkBear.find_links_in_file(file)` , to `InvalidLinkBear.find_links_in_file(self, file)`  makes sense in this commit and why changing it to self.find_links_in_file(file) deserves a separate commit?

Comment (SanketDG) [inline bears/general/InvalidLinkBear.py]: That makes sense, thanks!

**Comment by abhsag24** [comment]
ack  efeecde

File: bears/general/InvalidLinkBear.py [hunk 2]
```diff
@@ -8,15 +8,16 @@
 
 
 class InvalidLinkBear(LocalBear):
+    check_connection_url = "http://www.google.com"
```
Comment (Makman2) [inline bears/general/InvalidLinkBear.py]: Have you seen this: http://stackoverflow.com/questions/3764291/checking-network-connection  ? :)
> 74.125.228.100 is one of the IP-addresses for google.com. Change http://74.125.228.100 to whatever site can be expected to respond quickly. Using a numerical IP-address avoids a DNS lookup, which may block the urllib2.urlopen call for more than a second. Thanks to rzetterberg for pointing this out.

Comment (Makman2) [inline bears/general/InvalidLinkBear.py]: Though maybe we need to use a different IP, for me nothing happens :O

File: bears/general/InvalidLinkBear.py [hunk 3]
```diff
@@ -33,14 +44,14 @@ def find_links_in_file(file):
...
-                code = InvalidLinkBear.get_status_code_or_error(link)
+                code = self.get_status_code_checked(link)
                 yield line_number + 1, link, code
 
     def run(self, filename, file):
         '''
         Yields results for all invalid links in a file.
         '''
-        for line_number, link, code in InvalidLinkBear.find_links_in_file(file):
+        for line_number, link, code in (self.find_links_in_file(file)):
```
Comment (Makman2) [inline bears/general/InvalidLinkBear.py]: no need for the brackets :)

File: bears/general/InvalidLinkBear.py (hunk 2 above)
Comment (SanketDG) [inline bears/general/InvalidLinkBear.py]: Yeah its actually `216.58.218.174`. I'll add this and add a comment.

Comment (AbdealiJK) [inline bears/general/InvalidLinkBear.py]: In wondering here ...
Maybe we want a better check where we test for 8.8.8.8 and then check for google.com?
This way we can test if the dns server is not there or the internet connection itself is not there ?

-----Original Message-----

File: bears/general/InvalidLinkBear.py [hunk 4]
```diff
@@ -24,8 +25,18 @@ def get_status_code_or_error(url):
...
-    @staticmethod
-    def find_links_in_file(file):
+    def get_status_code_checked(self, url):
+        code = InvalidLinkBear.get_status_code(url)
+        if code is None:  # pragma: no cover
+            # if status code of google.com is None, then its
+            # an unstable connection.
+            if self.get_status_code(self.check_connection_url) is None:
+                self.err("Unstable connection detected. "
+                         "InvalidLinkBear will not continue.")
```
Comment (AbdealiJK) [inline bears/general/InvalidLinkBear.py]: Please use `self.__class__.__name__` so that we dont have to change this if the bear's name changed

File: bears/general/InvalidLinkBear.py [hunk 5]
```diff
@@ -24,8 +25,18 @@ def get_status_code_or_error(url):
...
             pass
 
-    @staticmethod
-    def find_links_in_file(file):
+    def get_status_code_checked(self, url):
+        code = InvalidLinkBear.get_status_code(url)
+        if code is None:  # pragma: no cover
+            # if status code of google.com is None, then its
+            # an unstable connection.
+            if self.get_status_code(self.check_connection_url) is None:
```
Comment (AbdealiJK) [inline bears/general/InvalidLinkBear.py]: `code is None and self.get_status_code(....`
nested ifs are basically an `and` as it's shortcircuited

**Comment by AbdealiJK** [comment]
ack c8f520d5fded99c15e9e44c1a64e7796e5a86217

File: bears/general/InvalidLinkBear.py (hunk 4 above)
Comment (SanketDG) [inline bears/general/InvalidLinkBear.py]: will do

**Comment by AbdealiJK** [comment]
@sils1297 just a note here.
I acked the last commit but not the second last one.
I remember you saying gitmate shouldnt set the PR status to OK in such cases. But it did set the PRs status to OK.
Bug ?

**Comment by abhsag24** [comment]
Same thing happened yesterday definitely a bug

-----Original Message-----

File: bears/general/InvalidLinkBear.py (hunk 5 above)
Comment (SanketDG) [inline bears/general/InvalidLinkBear.py]: I would like to keep it this way. This looks more concise and I don't have to wrap any lines.

File: bears/general/InvalidLinkBear.py (hunk 2 above)
Comment (Makman2) [inline bears/general/InvalidLinkBear.py]: if google works imo you have internet, no need to test for 8.8.8.8 too :)

File: bears/general/InvalidLinkBear.py (hunk 4 above)
Comment (Makman2) [inline bears/general/InvalidLinkBear.py]: then `type(self).__name__` :+1:

Comment (Makman2) [inline bears/general/InvalidLinkBear.py]: then `type(self).__name__` :+1:

File: bears/general/InvalidLinkBear.py (hunk 5 above)
Comment (Makman2) [inline bears/general/InvalidLinkBear.py]: I would agree with @AbdealiJK , no need to introduce another nesting level and using short-circuitance (I hope this is actually a word...) makes things slightly faster^^

**Comment by Makman2** [comment]
the github status check only checks the last commit on the PR. If you haven't acked the previous ones github doesn't recognize it^^

**Comment by AbdealiJK** [comment]
Sils had mentioned on gitter that now, with the new gitmate, after that huge down time it should check all the prs

-----Original Message-----

**Comment by sils1297** [comment]
gitmate doesn't track acks PR wide yet

Sincerely,

Lasse Schuirmann

contact@viperdev.io
http://viperdev.io/

2016-03-16 17:29 GMT+01:00 AbdealiJK <notifications@github.com>:

**Comment by sils1297** [comment]
@SanketDG any progress here? Need help?

**Comment by SanketDG** [comment]
@sils1297 Could this wait a few days? I'll get this done soon. I am not stuck, just prioritizing.

**Comment by sils1297** [comment]
sure no pressure, I was just checking up on some PRs :)

**Comment by sils1297** [comment]
ack https://github.com/coala-analyzer/coala-bears/pull/242/commits/2b6f689ec6baebb27a343a9f03aa6fe763aecc80

File: bears/general/InvalidLinkBear.py [hunk 6]
```diff
@@ -44,8 +57,7 @@ def run(self, filename, file, timeout: int=DEFAULT_TIMEOUT):
 
         :param timeout: Request timeout period.
         '''
-        for line_number, link, code in InvalidLinkBear.find_links_in_file(
-                file, timeout):
+        for line_number, link, code in self.find_links_in_file(file, timeout):
```
Comment (sils1297) [inline bears/general/InvalidLinkBear.py]: don't you want to try except here with the NoConnectionError

**Comment by sils1297** [comment]
I think https://github.com/coala-analyzer/coala-bears/pull/242/commits/9f66d38849a59e2af0abb831b5a09f0ebef9bbc6 needs work

File: bears/general/InvalidLinkBear.py (hunk 6 above)
Comment (SanketDG) [inline bears/general/InvalidLinkBear.py]: oh yes, should I `pass` in the except?

Comment (sils1297) [inline bears/general/InvalidLinkBear.py]: actually I think you'll rather want to do the `self.err` there and not in the get_status_code_checked function

**Comment by SanketDG** [comment]
reack ded1351

File: bears/general/InvalidLinkBear.py [hunk 7]
```diff
@@ -26,8 +29,16 @@ def get_status_code_or_error(url, timeout):
...
             pass
 
-    @staticmethod
-    def find_links_in_file(file, timeout):
+    def get_status_code_checked(self, url, timeout):
+        code = InvalidLinkBear.get_status_code(url, timeout)
+        # if status code of google.com is None, then its
+        # an unstable connection.
+        if code is None and self.get_status_code(
+                self.check_connection_url, timeout) is None:  # pragma: no cover
```
Comment (abhsag24) [inline bears/general/InvalidLinkBear.py]: why no cover? can't we stub and test?

File: bears/general/InvalidLinkBear.py [hunk 8]
```diff
@@ -26,8 +29,16 @@ def get_status_code_or_error(url, timeout):
         except requests.exceptions.RequestException:
             pass
 
-    @staticmethod
-    def find_links_in_file(file, timeout):
+    def get_status_code_checked(self, url, timeout):
+        code = InvalidLinkBear.get_status_code(url, timeout)
+        # if status code of google.com is None, then its
+        # an unstable connection.
+        if code is None and self.get_status_code(
```
Comment (abhsag24) [inline bears/general/InvalidLinkBear.py]: isn't `get_status_code` a `staticmethod`?

File: bears/general/InvalidLinkBear.py [hunk 9]
```diff
@@ -26,8 +29,16 @@ def get_status_code_or_error(url, timeout):
         except requests.exceptions.RequestException:
             pass
 
-    @staticmethod
-    def find_links_in_file(file, timeout):
+    def get_status_code_checked(self, url, timeout):
```
Comment (abhsag24) [inline bears/general/InvalidLinkBear.py]: i'm sorry if i missed something but i'm not sure why we need this function?, can't we have a `check_connection` param in the `get_status_code`  function and do the check there itself?

File: bears/general/InvalidLinkBear.py (hunk 8 above)
Comment (SanketDG) [inline bears/general/InvalidLinkBear.py]: thanks!

File: bears/general/InvalidLinkBear.py (hunk 9 above)
Comment (SanketDG) [inline bears/general/InvalidLinkBear.py]: the point of this function is to raise a error on the second try.
the point of `get_status_code` is to return the code.

Is that okay?

Comment (abhsag24) [inline bears/general/InvalidLinkBear.py]: why not something like
```
@staticmethod
def get_status_code(url, timeout, check_connection = True):
    try:
        code = requests.head(url, allow_redirects=False,
                            timeout=timeout).status_code
        if code is None and check_connection:
            if InvalidLinkBear.get_status_code(url, timeout,
                                              check_connection = False) is None:
                raise raise requests.exceptions.NoConnectionError
        return code
    except requests.exceptions.RequestException:
        pass
```

**Comment by sils1297** [comment]
This has conflicts. Is it still needed?

**Comment by Makman2** [comment]
@SanketDG ping :)

**Comment by SanketDG** [comment]
Closing this as per as #184
